File: Stacks/20. Valid Parentheses/20. Valid Parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        compliment = {")" : "(", "}": "{", "]": "["}
        stack = []
        
        for char in s:
            if char not in compliment:
                stack.append(c)
            else:
                if not stack or stack[-1] != compliment[c]:
                    return False
                stack.pop()
        
        return not stack

# Time: O(n)
# Space: O(n)

File: Stacks/20. Valid Parentheses/20. Valid Parentheses.js
/**
 * @param {string} s
 * @return {boolean}
 */
var isValid = function (s) {
  const compliment = { ")": "(", "}": "{", "]": "[" };
  let stack = [];

  for (let char of s) {
    if (!(char in compliment)) stack.push(char);
    else {
      if (stack.length === 0 || stack[stack.length - 1] !== compliment[char]) {
        return false;
      }
      stack.pop();
    }
  }

  return stack.length === 0;
};

// Time: O(n)
// Space: O(n)
